src/main/java/com/google/devtools/build/lib/syntax/Identifier.java

```diff
@@ -15,6 +15,7 @@
 package com.google.devtools.build.lib.syntax;
 
 import com.google.devtools.build.lib.util.SpellChecker;
+import com.google.devtools.build.lib.util.StringCanonicalizer;
```

Comment:
I haven't looked ahead in this stack of PRs yet, but:

Neither identifiers nor strings should be used as the key to an environment lookup, at least in an efficient (flat) representation. It is the job of the validation pass to relate each Identifier to the entity (in Skylark, always a variable) to which it refers, which is either a global, a local, a free variable, or a builtin. For each of those categories, and especially for the last two kinds, the validation pass also computes a dense numbering, so that variables can be identified by an index and accessed by array indexing. I suggest you follow the Go implementation:

https://github.com/google/skylark/blob/master/resolve/resolve.go
https://github.com/google/skylark/blob/master/syntax/syntax.go#L256

String interning is not relevant: equal but nonidentical strings (e.g. from different files) may denote the same variable; conversely, identical strings may denote different variables. Nor, I suspect, is it even beneficial, but of course that question is easily settled by profiling data.

Reply:
@alandonovan, you're correct that ideally an array should be used instead of dictionary, but I'm following https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/syntax/Identifier.java;l=23?q=Identifier.java which seems to be a good temporary compromise until we get to most efficient implementation based on arrays. Thanks for the pointers, btw!

Reply:
The linked comment says "for extra performance". Does this change actually result in extra performance? If not, what is the benefit of requiring that Identifier names be canonical?

Reply:
@alandonovan, as I've mentioned above, variable lookups have been called out as a bottleneck in a comment - https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/syntax/Identifier.java;l=27?q=Identifier.java

The improved performance in this case means specifically the performance of `==` vs `equals` which is obviously better. The question of whether it improves all use cases is a more complex one and so far I only tested this with only a few different scenarios where there does not seem to be a pronounced difference.

I would like to emphasize that this interning is not the main purpose of this PR, but just a tiny attempt for optimization. Since it clearly leads to some controversy, I'll remove this change and submit a separate PR with just this change to not derail the conversation from the main point of using `Identifier`s instead of `String`s.

Reply:
> The improved performance in this case means specifically the performance of == vs equals which is obviously better. 

The performance of == may be obviously better, but the performance of == combined with StringCanonicalizer is not obviously better or worse; nothing about it is obvious.

My whole point here may be summarized as: changes that purport to be optimizations must be justified by empirical data, not speculation. Program dynamics cannot be deduced from the source code alone.

Reply:
I totally agree with you, @alandonovan. Since my personal benchmarks showed only a marginal improvements (2-3%), pretty much the only reason why I made this change was just because it was very easy and I relied on the judgement of the comment author, but you're right that it's not the right way to approach optimizations :)

Reply:
> personal benchmarks showed only a marginal improvements (2-3%)

Now we're talking about empirical data---good. Could you elaborate on what you measured and what you found? A 2-3% improvement is no small thing.

I still don't understand why Identifier even needs an equals method.  Is it necessary and does its performance matter? Or is the observed performance gain a result of reducing persistent memory allocation?  (You could measure the effects independently by keeping StringCanonicalizer but reverting the Identifier.equals change.)

Reply:
Identifier is used as a key in a map (https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/syntax/LoadStatement.java) and a set (https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/syntax/LValue.java).

In both cases, it looks like we could use a list instead (we don't need fast lookup).